hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -924,128 +698,110 @@ def event_to_string(self, event):
+
+    def alphabet(self, n_bytes):
+        """Returns an alphabet - a list of values to use for all blocks with
+        this number of bytes - or None if this value should be generated
+        without an alphabet.
+
+        This is designed to promote duplication in the test case that would
+        otherwise happen with very low probability.
+        """
+        try:
+            return self.__alphabet[n_bytes]
+        except KeyError:
+            pass
+
+        if self.__random.random() <= self.pure_chance:
+            # Sometiems we don't want to use an alphabet (e.g. for generating
+            # sets of integers having a small alphabet is disastrous), so with
+            # some probability we want to generate choices that do not use the
+            # alphabet. As with other factors we set this probability globally
+            # across the whole choice of distribution so we have various levels
+            # of mixing.
+            result = None
         else:
-            c = rnd.randint(0, 255)
-        r.append(c)
-    return hbytes(r)
-
-
-class TargetSelector(object):
-    """Data structure for selecting targets to use for mutation.
-
-    The main purpose of the TargetSelector is to maintain a pool of "reasonably
-    useful" examples, while keeping the pool of bounded size.
-
-    In particular it ensures:
-
-    1. We only retain examples of the best status we've seen so far (not
-       counting INTERESTING, which is special).
-    2. We preferentially return examples we've never returned before when
-       select() is called.
-       If ``target()`` is in use, we maintain a pool of the best examples
-       for each known label, choose a random label, and choose an example
-       from that pool with higher-ranked examples more probable.
-    3. The number of retained examples is never more than the greater of
-       self.pool_size, with older examples discarded automatically, preferring
-       ones that we have already explored from.
-
-    These invariants are fairly heavily prone to change - they're not
-    especially well validated as being optimal, and are mostly just a decent
-    compromise between diversity and keeping the pool size bounded.
-    """
-
-    def __init__(self, random, pool_size=MUTATION_POOL_SIZE):
-        self.random = random
-        self.best_status = Status.OVERRUN
-        self.pool_size = pool_size
-        self.weights = tuple(accumulate(0.5 ** i for i in range(10)))
-        self.reset()
-
-    def __len__(self):
-        return (
-            len(self.fresh_examples)
-            + len(self.used_examples)
-            + sum(len(pool) for pool in self.scored_examples.values())
-        )
 
-    def reset(self):
-        self.fresh_examples = []
-        self.used_examples = []
-        self.scored_examples = defaultdict(list)
+            size = (
+                int(
+                    math.log(self.__random.random())
+                    / GenerationParameters.ALPHABET_FACTOR
```

Comment:
Might be good to have a direct explanation of what sort of random distribution this is trying to achieve.